### PR TITLE
Add Differential Equation API

### PR DESCRIPTION
**Pull request by Dpananos**
This work is a part of Google Summer of Code 2019.

This PR adds:

* `ode` submodule, which contains `DifferentialEquation`, the main tool for doing inference with ODEs.

* An `ODEGradop` for computation of the gradients required for HMC.

* A function for computing an augmented system of ODEs (that is, the original ODE + ODEs for the sensitivity equations) called `augment_system`.

* Unit tests for ode

* A Notebook showcasing how to use `DifferentialEquation` in a couple of examples regarding a free falling object and a spread of an infection.

* Rerun existing ODE notebook for comparison
---

Sorry, I was having some issues with the last PR so I thought I would just start again.[Here](https://github.com/pymc-devs/pymc3/pull/3578) is a link to the old PR should you need to reference it.

### CODE REVIEW

**Comment by michaelosthege** [comment]
Could the test failure be related to this issue? https://github.com/pymc-devs/pymc3/issues/1640

**Comment by Dpananos** [comment]
FYI: Test pass now because I have included the `pytest.xfail` decorator on previously failing tests.  See `pymc3/tests/test_ode.py`

**Comment by ColCarroll** [comment]
I think that's a good choice. For reference, this means it might not work on GPUs -- I am not sure if that is an OK tradeoff for ODEs, but I *do* think such a failure should not block this PR. We should remember to open an issue once it is merged, though.

File: pymc3/ode/ode.py
Comment (ColCarroll) [inline pymc3/ode/ode.py]: add a newline here

File: pymc3/ode/utils.py
Comment (ColCarroll) [inline pymc3/ode/utils.py]: this is a guess, but could you provide defaults here to avoid the `test_value` errors you have been seeing? seems like while instantiating all these vectors (`t_y`, `t_p`, `t_t`, `dydp_vec`), you know the shape, so you can initialize them to all zeros?

File: pymc3/tests/test_ode.py
Comment (ColCarroll) [inline pymc3/tests/test_ode.py]: i might be the only one who gets annoyed here, so please ignore if it is trouble, but in function calls, we should not have spaces around the `=` sign:
```python
DifferentialEquation(func=1,
                     t0=0,
                     times=self.times,...
)
```

File: pymc3/tests/test_ode.py
Comment (ColCarroll) [inline pymc3/tests/test_ode.py]: ...is this `black`'s doing? i might make it one line otherwise

File: pymc3/ode/ode.py
Comment (ColCarroll) [inline pymc3/ode/ode.py]: delete the extra spaces around `parameters[...]`

File: pymc3/ode/ode.py
Comment (ColCarroll) [inline pymc3/ode/ode.py]: i'm not sure what the intention is here, but this construction looks tricky. I find `args=(parameters,)` to be the most readable if that is what you're looking for:

image

**Review by ColCarroll — COMMENTED** [review]
Left a few nitpicky things, and one musing suggestion. None of them are dealbreakers, and I will be delighted when this gets merged, even if none of the changes get made.

As another musing suggestion, running 

```bash
python -m pylint --enable=all pymc3/step_methods/hmc/nuts.py
```

on each file you changed  -- many suggestions will be silly, or ones we do not enforce, but some are helpful.

If you want to be really fancy

```bash
git diff --name-only HEAD master | xargs python -m pylint --enable=all
```

will pipe just the changed files through pylint

File: pymc3/tests/test_ode.py
Comment (Dpananos) [reply pymc3/tests/test_ode.py]: Lol, no that is a perfectly fine knitpick.  That is an artifact of my brain trying to cope with spaces after assignment but not after arguments.

File: pymc3/tests/test_ode.py
Comment (Dpananos) [reply pymc3/tests/test_ode.py]: I think that is VS code trying to be helpful.

**Comment by Dpananos** [comment]
Thanks @ColCarroll, I think I added everything you suggested.

File: pymc3/ode/ode.py
Comment (michaelosthege) [inline pymc3/ode/ode.py]: typos in both ValueErrors

File: pymc3/ode/ode.py
Comment (michaelosthege) [inline pymc3/ode/ode.py]: The props should not be empty. You can set it to `'func', 't0', 'times', 'n_states', 'n_odeparams'`.

File: pymc3/ode/ode.py
Comment (michaelosthege) [inline pymc3/ode/ode.py]: Can this line be dropped with your recent introduction of test values?

File: pymc3/ode/ode.py
Comment (michaelosthege) [inline pymc3/ode/ode.py:142]: please add a comment what this reshape + dot product do

**Comment by michaelosthege** [comment]
Your module is currently not imported by default. Please add this after line 10 of `pymc3\__init__.py`:
```
from . import ode
```
(just like `gp` is imported in line 7)

File: pymc3/ode/__init__.py
Comment (twiecki) [inline pymc3/ode/__init__.py]: missing new line

File: pymc3/ode/ode.py
Comment (twiecki) [inline pymc3/ode/ode.py]: should be relative imports

File: pymc3/ode/ode.py
Comment (twiecki) [inline pymc3/ode/ode.py]: no new-line (I think)

File: pymc3/ode/ode.py
Comment (twiecki) [inline pymc3/ode/ode.py]: please make code pep8 compatible, no spaces around kwargs, but spaces around math, space after comment

**Comment by twiecki** [comment]
As this is new code, can you just run black on the new files?

File: pymc3/ode/__init__.py
Comment (Dpananos) [reply pymc3/ode/__init__.py]: After `from .ode import DifferentialEquation`?  Just add a blank line?

File: pymc3/ode/ode.py
Comment (Dpananos) [reply pymc3/ode/ode.py]: Done.

File: pymc3/ode/ode.py
Comment (Dpananos) [reply pymc3/ode/ode.py]: Done

File: pymc3/ode/ode.py
Comment (Dpananos) [reply pymc3/ode/ode.py]: files in `pymc3.gp` use fixed imports. See [here](https://github.com/pymc-devs/pymc3/blob/master/pymc3/gp/gp.py#L8).  Want me to change that in another PR?

**Comment by Dpananos** [comment]
@twiecki Ran black on all the new files.

**Review by michaelosthege — APPROVED** [review]
@Dpananos running `black` gave some nice results there! Can you also run it on `test_ode.py`?

@twiecki @ColCarroll Are we ready to merge?

There are some improvements regarding the returned shape & the `grad` that will follow in the near future, but the functionality is already there with this PR.

File: pymc3/ode/ode.py
Comment (twiecki) [inline pymc3/ode/ode.py:33]: the style fixes are not applied here

**Comment by twiecki** [comment]
It seems like black was run on just ode.py, can you run it on all new files here?

**Comment by Dpananos** [comment]
I ran black on the `pymc3/ode` directory.  It left the majority of the files unchanged.  I made the change in the docstring, and also ran black on the `test_ode.py` file.

File: pymc3/ode/ode.py
Comment (canyon289) [inline pymc3/ode/ode.py]: Is a triple quote docstring needed on this method?

File: pymc3/ode/ode.py
Comment (canyon289) [inline pymc3/ode/ode.py]: I believe this should be formatted as such
```
def _system(xyx)
     """This is the function that will be passed to odeint.
...
```

**Comment by canyon289** [comment]
Just some minor comments on some code style. The notebook could also use spaces on the inline comments. I know its picky, but for me its jarring after seeing a bunch of PEP8 style comments.

On the notebook the PyMC3 version is shown, just to double check will that be the version of PyMC3 that includes the ODE solver, or will we increment the PyMC3 version to 3.8 once this ODE solver get merged to master?

File: pymc3/ode/ode.py
Comment (Dpananos) [reply pymc3/ode/ode.py]: This is a helper function to create an initial condition for the augmented differential equation.  I'll add one anyway because, hey, what's the harm?

Comment (twiecki) [reply pymc3/ode/ode.py]: @canyon289 I think this is all black.

File: pymc3/ode/ode.py
Comment (twiecki) [reply pymc3/ode/ode.py]: Again, black (I think).

**Comment by twiecki** [comment]
Need to add to release-notes and link the new NBs to the docs.

File: pymc3/ode/ode.py
Comment (canyon289) [reply pymc3/ode/ode.py]: @twiecki I don't think black will get this one because it requires a change from a single line docstring to triple quite

@Dpananos Thanks for making the change. Even if its a helper function someone might have to read it someday and being able to introspect docs through the normal tools cant hurt

File: RELEASE-NOTES.md
Comment (twiecki) [inline RELEASE-NOTES.md]: Can you move this to the top? That will be the highlight feature of this release.

**Comment by Dpananos** [comment]
@twiecki Not sure how to update the docs, but I did find a reference to the other ODE notebook in ` docs/source/notebooks/table_of_contents_examples.js`.  I just swapped out my notebook for that one.

@canyon289 Changes made and made the notebook more pep-8 friendly wrt spaces (I will never get used to one space after assignment, no space after arguments).

If that's all, are we ready to merge?

File: RELEASE-NOTES.md
Comment (Dpananos) [reply RELEASE-NOTES.md]: Oops, yes absolutely.

File: RELEASE-NOTES.md
Comment (twiecki) [inline RELEASE-NOTES.md]: Can you link to this PR?

File: docs/source/notebooks/table_of_contents_examples.js
Comment (twiecki) [inline docs/source/notebooks/table_of_contents_examples.js:56]: If we are unlinking this NB (should we actually do that?) we should probably also remove the NB.

File: pymc3/ode/ode.py
Comment (twiecki) [reply pymc3/ode/ode.py:33]:
```suggestion
            # Logistic differential equation
```

File: pymc3/ode/ode.py
Comment (twiecki) [inline pymc3/ode/ode.py]:
```suggestion
            return p[0] * y[0] * (1 - y[0])
```

File: pymc3/ode/ode.py
Comment (twiecki) [inline pymc3/ode/ode.py]:
```suggestion
        times = np.arange(0.5, 5, 0.5)
```

File: pymc3/ode/ode.py
Comment (twiecki) [inline pymc3/ode/ode.py]: Elsewhere we follow a different style for doc strings, I believe it's what numpy uses, can you check other parts of the code-base? Did you compile the docs? I don't think this will show up correctly.

Comment (fonnesbeck) [reply pymc3/ode/ode.py]: I assume this is a private method, given that it is named with an underscore? If so, you might remove the args description entirely and just leave the comment.

Comment (rpgoldman) [reply pymc3/ode/ode.py]: > Elsewhere we follow a different style for doc strings, I believe it's what numpy uses, can you check other parts of the code-base? Did you compile the docs? I don't think this will show up correctly. [@twiecki ]

Yes, our docs are written in numpy style.  If you Google for instructions, the style guide is easy to find, and docs written in numpy style will be properly processed by our sphinx configuration so that they appear on the website.f

File: docs/source/notebooks/table_of_contents_examples.js
Comment (Dpananos) [reply docs/source/notebooks/table_of_contents_examples.js:56]: No, let's not unlink it or delete it.  I'll just add another line for my notebook and keep things as they are.  The original notebook is still useful IMO

File: RELEASE-NOTES.md
Comment (Dpananos) [reply RELEASE-NOTES.md]: Done.

File: pymc3/ode/utils.py
Comment (twiecki) [inline pymc3/ode/utils.py:14]: These still have the old doc-format.

Comment (twiecki) [reply pymc3/ode/utils.py:14]: Merged and fixed on master.

**Comment by twiecki** [comment]
Congrats again @Dpananos, this was a beast!

**Pull request merged**
